### ISSUE TITLE
Local Studio - Feature Request for Saving Snippets Locally

### ISSUE THREAD

**Issue by rcwestlake** [supabase/supabase#17941]
# Bug report

## Describe the bug
In the local studio dashboard, I cannot edit a SQL Query name

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. In the local studio, go to the SQL editor
2. Click "New query"
3. Do not edit the name, click elsewhere
4. Come back, you can no longer edit the name and it says "Untitled query"

## Expected behavior
I should be able to edit the query name

## Screenshots

![image](https://github.com/supabase/supabase/assets/9679076/ce368bd4-0ce0-4b41-91f7-9226eb27553e)

## System information

- OS: mac
- Browser (if applies): Brave
- Version of supabase-js:  @supabase/supabase-js": "^2.36.0
- Version of Node.js: v20.5.1

**Comment by saltcod** [comment]
Thanks @rcwestlake — not able to reproduce this one. Still happening?

**Comment by whoiscarlo** [comment]
I'm experiencing the same thing with version 2.38.1 on Arc Browser, Safari, and Chrome. The dropdown menu that's typically on the right side of the item label is no where to be found.

**Comment by pierre-aurele-martin** [comment]
I can also confirm this issue. 
`supabase/studio:20230921-d657f29` 


I don't see any related error in console :)

**Comment by whoiscarlo** [comment]
Also, whenever do you make a query, close the window and reopen SQL editor all created queries are gone and only the defaults remain.

**Comment by sonnyvesali** [comment]
any update on this I'm using supabase-cli version 1.110.1

**Comment by meenie** [comment]
This is still happening on 1.113.3. Can we get an update on a possible fix?

**Comment by yusrilmohammad** [comment]
confirm bug with these version:
**supabase/studio:20231030-9abc686**
**supabase/studio:20231123-64a766a**
**supabase/studio:20231127-81e146f**

![g492](https://github.com/supabase/supabase/assets/114605873/33de676b-cfc1-4230-837e-174b29a473ce)
method **PUT** not allowed on endpoint: api/projects/default/content

but when i used older image:
**supabase/studio:20221214-4eecc99**

its works and its used method **PATCH**
![g554](https://github.com/supabase/supabase/assets/114605873/4fe5adf4-604c-4fd3-9273-639b9816ab1b)

**Comment by davhin** [comment]
had the same issue on web just now, logging out and back in fixed that for me

**Comment by levibe** [comment]
> had the same issue on web just now, logging out and back in fixed that for me

Same here. Dropdown was showing but `Rename query` action was not in it. Logging out and back in fixed it.

**Comment by santhonydo** [comment]
can confirm issue still exist on studio. i'm on 1.131.4

**Comment by encima** [comment]
Hi folks, thanks for reporting back on this. We are looking into this now and will triage it to the frontend team!

**Comment by saltcod** [comment]
Hi all — apologies for the delay on this one.  

**Quick update:** 
For hosted (supabase.com) projects: This was a bug  that should be resolved now. 

For local and self-hosted projects, there currently isn't a way to save queries to your project. Hence we just give you one SQL workspace to work from, without the concept of individual snippets. We recognize that this isn't ideal and are thinking of solutions to allow you to save snippets here like you do on supabase.com.  

We appreciate the reports and hope to get a solution out soon.

**Comment by whoiscarlo** [comment]
@saltcod @encima would you be able to just store the queries into a text file of some kind into the supabase folder that's created and then just read from it? Wouldn't need to be anything fancy =)

**Comment by encima** [comment]
> @saltcod @encima would you be able to just store the queries into a text file of some kind into the supabase folder that's created and then just read from it? Wouldn't need to be anything fancy =)

Hi @whoiscarlo,

I like your creativity. Unfortunately, it is not so simple as it is harder to maintain and the functionality to do that is quite different from the platform. We are discussing this internally and will update as soon as there is one!

**Comment by olee** [comment]
> For local and self-hosted projects, there currently isn't a way to save queries to your project. Hence we just give you one SQL workspace to work from, without the concept of individual snippets. We recognize that this isn't ideal and are thinking of solutions to allow you to save snippets here like you do on supabase.com.

Well - how about just some simple local-storage fallback at least? Wouldn't be some great solution but at least serve it's purpose quite well for now.
But loosing all written SQL just by maybe an accidential reload is quite annoying tbh.

**Comment by joshenlim** [comment]
just FYI i'm gonna change the tag from bug to enhancement, since this isn't a bug 🙏

**Comment by LazerJesus** [comment]
+1 please

**Comment by olee** [comment]
Bump.
Considering this could be quite the quick-win, any idea on when this might be fixed?

**Comment by encima** [comment]
It is still being considered how this can best be done, we will update here as soon as we have more info.

**Comment by patrickdundas** [comment]
This is very basic UX functionality, please fix this bug

**Comment by encima** [comment]
Apologies for the silence here, it is being considered still. The problem of implementing a "quick fix" is that it needs to continue to be supported so, ideally, we want the self-hosted Studio to be as close to the hosted version as possible.

That said, we are very open to PRs so feel free @olee (the code is in the SQL editor code under the `apps/studio` folder)

**Comment by dfang** [comment]
confirm bug with the latest version:

```
image: supabase/studio:20240514-6f5cabd
```

**Comment by parkernilson** [comment]
I would absolutely love this feature!!

**Comment by nilsertle** [comment]
Since this feature is still not implemented, does anyone have a better workflow for local development? I am not an expert in databases and was wondering if I can also develop from local SQL files instead of using the SQL editor in the supabase studio? Losing all the written code has really caused some frustration.

**Comment by ptonewreckin** [comment]
Any updates on this?

**Comment by encima** [comment]
> Since this feature is still not implemented, does anyone have a better workflow for local development? I am not an expert in databases and was wondering if I can also develop from local SQL files instead of using the SQL editor in the supabase studio? Losing all the written code has really caused some frustration.

Yes, 100%! You can put code in your `supabase` folder and apply it through `migrations`.

Check out [migrations](https://supabase.com/docs/reference/cli/supabase-migration) for this.

Another option would be to apply `sql` files through `psql`

**Comment by adao** [comment]
At the last could show the actual queries with the text clipped. As of now, it's just rows of 'Untitled query'

**Comment by nilsertle** [comment]
> > Since this feature is still not implemented, does anyone have a better workflow for local development? I am not an expert in databases and was wondering if I can also develop from local SQL files instead of using the SQL editor in the supabase studio? Losing all the written code has really caused some frustration.
> 
> Yes, 100%! You can put code in your `supabase` folder and apply it through `migrations`.
> 
> Check out [migrations](https://supabase.com/docs/reference/cli/supabase-migration) for this.
> 
> Another option would be to apply `sql` files through `psql`

This solution has been helpful. I have created a file structure to organize all my logic, making it much more readable. However, it remains somewhat messy due to the fact that it is pure SQL and not a backend framework. This becomes noticeable when you need to use global variables or when your logic becomes more complex, as it can reduce readability.
Is there perhaps a better linter for PostgreSQL that can visually represent the files in a more organized manner? Alternatively, is there a way to use a backend in front of the SQL code? The problem I face is that I do not want to host an additional backend and incur the associated costs. Supabase is free in the initial stages, which is why I am currently using it. I would really like to here some best practices. Maybe there is a source where I can do some more research. My architecture knowledge is very limited.

**Comment by parkernilson** [comment]
@nilsertle I have had a much better experience using Datagrip IDE. It has a full-fledged language server for Postgres, and makes the developer experience of working with Postgres SO much better. The caveat is that it is a paid service (although I get it free through my university).

There are also several open source projects that are in the works to implement a similar experience to that of Datagrip in other editors like Neovim and VSCode. The most notable being [Supabase Postgres LSP](https://github.com/supabase-community/postgres_lsp). Once there is an open source full LSP for Postgres development will be a lot more manageable!

**Comment by ErwinAI** [comment]
> > how about just some simple local-storage fallback at least
> 
> We did consider this! More specifically using index DB - in the midst of exploring options
> 
> separately just FYI i'm gonna change the tag from bug to enhancement, since this isn't a bug 🙏

With all due respect - but I do think that this is a bug, not an enhancement. The UI shows "Untitled query" when clicking "new query". Both the appearing of the untitled query in the menu, as well as the button 'new query' imply that:

1) Queries can be new, which doesn't make any sense if they cannot be saved.
2) Queries can be untitled, which implies that they can be titled as well. This requires saving, which is currently not possible.

I don't think it makes logical sense to label this as an enhancement. It is, derived from the above logic, a bug.

In two months time, this issue is 1 year old. Is there any progress or decisions?

The nudge to contribute is helpful, @encima, but it isn't very appealing to invest time when we don't know the _internal decision(s)_ regarding this.

**Comment by encima** [comment]
Agreed and we do apologise for this. The self-hosted functionality is a key reason for why I chose to work at Supabase and the work that is being done on the self-hosted side by the team is amazing. As an anxious person (aren't we all?), seeing a notification from this Issue strikes me with fear and sadness because I know you all want it solved and I also know how hard the @supabase/frontend team is working

Keeping these features in sync is tough and, when the self-hosted version does not have a management API, it becomes a real challenge. **Personal opinion: I would like to see a dummy management API in the self-hosted version that could mimic some of the functionality.**

Of course, as many people here have suggested alternative approaches, we do not want 3 different PRs opened with different approaches.

So..my suggestion is this:
- I organise a meeting with the Frontend team to discuss what could be seen as an acceptable implementation and post the notes from that meeting here
- I will create a branch with an attempt to start this implementation (sadly, the frontend is not written in Python in 2010 so it may suck)
- Those willing can then contribute to this branch so that we know effort is combined and the solution is not wasted
- If the frontend team agrees, we can aim to implement this (community led) before this issue is 1 year old and the team can then review it.

Thoughts and alternatives welcome!

**Comment by ErwinAI** [comment]
@encima Thank you for the prompt response, and understandable that it makes invokes some feelings seeing a notification from this issue (first hand experience with that.. ). It's all good, just wanted to sharpen the situation with this issue and move it towards completion seeing the amount of people that run into this. First step in my head was changing the label back to 'bug' to increase urgency.

The approach you listed is very welcome, that'll probably move it towards completion rapidly.

It might also be a good idea to open an issue for implementing a dummy version of the management API and do a call for collaboration on that?

In any case, I'll keep my eyes on this issue and if help is needed with implementation I'll see if I can help.

**Comment by encima** [comment]
Hi @ErwinAI 

Thanks for this! Rest assured that this has not been forgotten and it is an important issue. Solving it with a dummy management API is the approach we like internally as it allows us to scaffold other features and brings tighter parity between hosted and self-hosted.

Sadly, it is possibly the larger task! 
The pros of this are:
- Our APIs are all following the OpenAPI spec and available online or in repos
- There are tools to create dummy servers/proxies from these specs (i.e. [Stoplight Prisma](https://stoplight.io/open-source/prism))

The cons are:
- Requires time to investigate and might be a large investment -> longer time to implement
- There is no tool I am aware of that does something similar to Stoplight Prisma AND interacts with a datastore (though they do support custom functions)

When we have more info, I will report back here!

**Comment by yekta** [comment]
I've been waiting for this a long time, more than a year at this point (before this issue and before my own issue). I think even something like local storage would've saved quite a bit of headache for self-hosted users.

We've been saving our queries to notes, which feels ironic considering you can do so much with Studio, this one feels much less complicated and a basic feature while very useful at the same time. It feels pretty weird to see your queries not being saved while you are using a management system for a database since you have an entire database used by that system to save those queries to 😅 I'm aware that's not the scope of the queries in the hosted version (we've used it for a long time). We can get suggestions from a state of the art LLM to edit or create queries, yet we can't save them, while operating on a database.

That being said, I do understand how giving self-hosted users a way to save their queries might rank very low in the list of priorities for Supabase considering how many other things Supabase has been doing since the start.

**Comment by cyluss** [comment]
@encima Dummy management API is impressive idea. I appreciate your enthusiasm and I hope to contribute in the future

**Comment by ZhenFTW** [comment]
hello?

**Comment by encima** [comment]
@ZhenFTW 👋

**Comment by andersoncampolina** [comment]
any news? I want to be my own cloud serviceeeeee :p

**Comment by Vixy88** [comment]
@encima did you create a branch for this to try and solve? :)

**Comment by encima** [comment]
@Vixy88 I have a proxy of the management API and am waiting on the frontend team to determine the next steps

**Comment by Vixy88** [comment]
> @Vixy88 I have a proxy of the management API and am waiting on the frontend team to determine the next steps

Okay great, let me know if you want any help.

Sick of saving my snippets as functions..:D

**Comment by Shabinder** [comment]
> Sick of saving my snippets as functions..:D

+1,

Open to contribute on this, let us know with any update on same, thanks 🙌

**Comment by oiqm4T** [comment]
waiting for fix, thanks for all the hard work!

**Comment by moatus** [comment]
Still an issue with local deployments. Quite discouraging that it's been so long.

**Comment by pbsullOptima** [comment]
+1

Also, any reason these can't be persisted in a protected schema like Auth/Storage/Vault? Then it can be saved locally?

**Comment by carlosfarfangalindoNL222** [comment]
Please fix this, is really a need for self hosting!

**Comment by wost-digital-solutions** [comment]
confirm bug with this version: 20240923-2e3e90c

**Comment by ChuckJonas** [comment]
I keep trying to run queries with `cmd-r` and reloading my page lol... Please give us a way to save things!  Even if it's just local storage based

**Comment by tvogel** [comment]
For people who are hampered by the lack of locally saved snippets, a workaround that I quite like:
Use a Jupyter/IPython notebook (e.g. in VS Code) and connect to your supabase using [JupySQL](https://jupysql.ploomber.io/en/latest/integrations/postgres-connect.html) (of course you don't need to spin up PostgreSQL as in the doc - use your supabase).

**Comment by cvillagrasa** [comment]
Thanks a lot for this suggestion, @tvogel ! That can work very well in the meantime.

I was trying it out and I'm finding that with everything default, the following Jupyter cell gives me a "Tenant or user not found" error when using JupySQL:

```
%sql postgresql://postgres:your-super-secret-and-long-postgres-password@localhost:5432/postgres
```

Up to now, I had never worked with a direct PostgreSQL connection on a self hosted Supabase. Nmap shows localhost:5432 is open.

Can you spot any obvious flaw?

**Comment by cvillagrasa** [comment]
Ok, it seems the database url needs to include the tenant id, via https://github.com/supabase/supabase/issues/29892#issuecomment-2419319879

Correct line:
```
%sql postgresql://postgres.your-tenant-id:your-super-secret-and-long-postgres-password@localhost:5432/postgres
```

**Comment by tvogel** [comment]
> Ok, it seems the database url needs to include the tenant id, via [#29892 (comment)](https://github.com/supabase/supabase/issues/29892#issuecomment-2419319879)
> 
> Correct line:
> 
> ```
> %sql postgresql://postgres.your-tenant-id:your-super-secret-and-long-postgres-password@localhost:5432/postgres
> ```

Could this be the Supavisor connection pooler port? Here, PostgreSQL runs on 54322. If you run `npx supabase status`, it should give you the DB URL.

**Comment by cvillagrasa** [comment]
I didn't set it up from the Supabase CLI, but from running docker compose:
[https://supabase.com/docs/guides/self-hosting/docker](https://supabase.com/docs/guides/self-hosting/docker)

That should make the difference in ports. Anyway, I can confirm that JupySQL works very well as an interactive Postgres tool with code persistence. Perfect workaround for the moment.

**Comment by volfadar** [comment]
up, it's been more than 1 year :(

**Comment by carlosromanxyz** [comment]
It's my first time using Supabase, on web works fine, but on local I've same issue with DELETE method for auth.users (405 error). Currently I'm working only with this table, I'm not try another table.

I will watch this issue. One year, omg...

(Excuse my english)

**Comment by zxkevinxz** [comment]
Update?

**Comment by jorisebbelaar** [comment]
Being able to save the queries on the self hosted version would be extremely helpfull

**Comment by mkarkee** [comment]
Hopefully this gets fixed soon.

**Comment by romek-rozen** [comment]
Do you have plan to fixing that?

**Comment by castortech** [comment]
> Hi all — apologies for the delay on this one.
> 
> **Quick update:** For hosted (supabase.com) projects: This was a bug that should be resolved now.
> 
> For local and self-hosted projects, there currently isn't a way to save queries to your project. Hence we just give you one SQL workspace to work from, without the concept of individual snippets. We recognize that this isn't ideal and are thinking of solutions to allow you to save snippets here like you do on supabase.com.
> 
> We appreciate the reports and hope to get a solution out soon.

Have you had time to cogitate enough and come to a solution ?

**Comment by SahilAgarwal321** [comment]
@saltcod Cant you just use the docker volume to store them? That would be an easy enough implementation as well, as you already use docker volume to store data.

**Comment by tvogel** [comment]
After having worked initially with the SQL snippets on hosted (nano) supabase to define all my DB logic and now for a while with local dev (`supabase start`) and self-hosted using Jupyter notebooks with JupySQL and migration files, I can only recommend not to use snippets: They are not versioned, they are hard to maintain - honestly, while it is cool for a quick start, for me, it is a bit of an anti-pattern.
If you are frustrated by the lack of it, use it to your advantage and make Jupyter notebooks that can also have better documentation and are easy to version control and also easy to share with the team.

**Comment by bradennapier** [comment]
i mean you have supabase storage :-P just create a internal / protected bucket and save to that :-P

**Comment by carsonbird** [comment]
Also experiencing this.

**Comment by kallebysantos** [comment]
Hi @carsonbird 💚
Currently this feature is not implemented for self-host/local-dev 😕

Only available on platform

**Comment by techrac** [comment]
being able to save queries on the local instance seems to be a reasonable request. Especially because its work in production supabase, maybe its time to get this one sorted ? would be very much appreciated, thanks!

**Comment by aantti** [comment]
There's an ongoing PR for that :) Appreciate your patience. Hopefully we'll have it soon.

**Comment by ivasilov** [comment]
We've just released a solution for this issue, you can take a look at https://github.com/orgs/supabase/discussions/42031. Please let us know if you run into any issues.
